ADR-51: Document DiscardNew interaction (#419)
nats-io/nats-architecture-and-design: adr/ADR-51.md

| `Nats-Schedule`           | The schedule the message will be published on                                                                                                                                                                                                                                                                                                                                                                          |
| `Nats-Schedule-Target`    | The subject the message will be delivered to                                                                                                                                                                                                                                                                                                                                                                           |

> | `Nats-Schedule-Source`    | Instructs the schedule to read the last message on the given subject and publish it to the target. If no message exists on the source subject, the schedule's own body and headers is published as a fallback and an advisory is emitted with reason `source_empty` (a second fallback trigger exists under `DiscardNew`, see [Discard Policy Interaction](#discard-policy-interaction)). Wildcards are not supported. |

Reviewer: bit worried about the amount of advisories here, schedules can fire every second - and I think for sampled ones would be on the more frequent side - so this could result in a lot of advisories.  We already publish an empty message on such a schedule and that might be enough?

Reviewer: Advisories are for exceptional circumstances and audit trails, I dont think this situation really is either of those.  And those who get advisories are not going to care for this information its a app concern not a system concern

Author: Removed the advisories.

Option 2 is generally the more robust configuration because schedule persistence no longer depends on a consumer that exists solely to hold interest.

#### Discard Policy Interaction

The discard policy is orthogonal to the retention policy above. The retention table covers the eviction-based limits (`DiscardOld`, `MaxAge`, `MaxMsgs`, `MaxBytes`), all of which can remove a stored schedule and disable it. `DiscardNew` is the opposite mechanism: instead of evicting an existing message to make room, it rejects the incoming write. As a result, a schedule message, once stored, is **protected** from `MaxMsgs`/`MaxBytes` eviction under `DiscardNew`. `MaxAge` is unaffected by the discard policy and always expires the schedule by age, so the `Nats-TTL` recommendation above still applies.

Two distinct writes are affected when the stream is full (`MaxMsgs`/`MaxBytes`, or `MaxMsgsPerSubject` under `DiscardNewPerSubject`):

> - **Storing the schedule.** Publishing a new schedule onto a fresh subject is a net-new message and is rejected when the stream is full, the same as any other publish. Updating an existing schedule on the same subject is a rollup (`Nats-Rollup: sub`) that replaces the prior message rather than growing the count. Under `MaxMsgs` this is net-zero on message count and so reliably succeeds even at the limit. Under `MaxBytes` the purge frees the old message's bytes, but if the new schedule is larger such that the result would exceed `MaxBytes` it is still rejected. Both `MaxMsgs` and `MaxBytes` are absolute limits that are not allowed to be exceeded.

Reviewer 2: ```suggestion
- **Storing the schedule.** Publishing a new schedule onto a fresh subject is a net-new message and is rejected when the stream is full, the same as any other publish. Updating an existing schedule on the same subject is a rollup (`Nats-Rollup: sub`) that replaces the prior message rather than growing the count. Under `MaxMsgs` this is net-zero on message count and so reliably succeeds even at the limit. Under `MaxBytes` the purge frees the old message's bytes, unless the new schedule is larger such that the result would exceed `MaxBytes`, in which case it is rejected. Both `MaxMsgs` and `MaxBytes` are absolute limits that are not allowed to be exceeded.
```

  - **Repeating schedules (cron and `@every`)**: the schedule message must remain to keep firing, so each trigger is a net-new message. When the stream is full the generated message is rejected and dropped for that tick, exactly like any other producer hitting `DiscardNew`. The schedule message is untouched and the schedule is **not** disabled; it fires again on the next tick and succeeds once space frees up (via `MaxAge` expiry, a consumer draining the stream, or an explicit purge/delete). The server does not stall, queue, or back-fill skipped scheduled messages.

For reliable schedule replacement prefer `MaxMsgs`: a replace is always net-zero on message count, whereas under `MaxBytes` a larger replacement can still be rejected.

A schedule that does not produce its intended message emits an advisory so the outcome is not silent. This covers a repeating schedule that is skipped, as well as a source-backed schedule that falls back to the schedule's own content:

```

> subject: $JS.EVENT.ADVISORY.STREAM.SCHEDULE.DEGRADED.<stream>

Reviewer: as above, dont think we need advisories here
